src/tactic/dependencies.lean

```diff
@@ -0,0 +1,230 @@
+/-
+Copyright (c) 2020 Jannis Limperg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jannis Limperg
+-/
+import data.list.defs
+import meta.expr
+import tactic.core
+
+/-!
+# Tactics about dependencies
+
+An expression `e` depends on another expression `f` if `f` occurs in `e` (see
+note [dependencies of hypotheses] for details and corner cases). This module
+provides tactics for checking whether hypotheses depend on each other and for
+enumerating the dependencies and reverse dependencies of hypotheses.
+-/
+
+namespace tactic
+
+/--
+Given a hypothesis (local constant) `h` and a local constant `i`, we say that
+`h` depends on `i` iff
+
+- `i` appears in the type of `h` or
+- `h` is a local definition and `i` appears in its body.
+
+We say that `h` inclusively depends on `i` iff `h` depends on `i` or `h = i`
+(so inclusive dependency is the reflexive closure of regular dependency).
+
+For example, consider the following context:
+
+```lean
+P : ∀ n, fin n → Prop
+n : ℕ
+m : ℕ := n
+f : fin m
+h : P m f
+```
+
+Here, `m` depends on `n`; `f` depends on `m`; `h` depends on `P`, `m` and `f`.
+Note that `f` and `h` do not depend on `n`, so the depends-on relation is not
+transitive in the presence of local definitions. `h` inclusively depends on `h`,
+`P`, `m` and `f`.
+
+We also say that `m` is a dependency of `f` and `f` a reverse dependency of `m`.
+Note that the Lean standard library sometimes uses these terms differently:
+`kdependencies`, confusingly, computes the reverse dependencies of an
+expression.
+-/
+library_note "dependencies of hypotheses"
+
+
+/--
+`type_has_local_in h ns` returns true iff the type of `h` contains a local
+constant whose unique name appears in `ns`.
+-/
+meta def type_has_local_in (h : expr) (ns : name_set) : tactic bool := do
+  h_type ← infer_type h,
+  pure $ h_type.has_local_in ns
+
+/--
+`local_def_value_has_local_in h ns` returns true iff `h` is a local definition
+whose body contains a local constant whose unique name appears in `ns`.
+-/
+meta def local_def_value_has_local_in (h : expr) (ns : name_set) : tactic bool := do
+  (some h_val) ← try_core $ local_def_value h | pure ff,
+  pure $ h_val.has_local_in ns
+
+/--
+Given a hypothesis `h`, `hyp_depends_on_locals h ns` returns true iff `h`
+depends on a local constant whose unique name appears in `ns`. See note
```

Comment:
```suggestion
depends on a local constant whose unique name appears in `ns`. See Note
```
and below. I think the regex parser in doc gen only looks for capital `Note [...]`.